XenAdmin/Wizards/PatchingWizard/PatchingWizard_SelectPatchPage.cs

```diff
@@ -179,14 +178,23 @@ public WizardMode WizardMode
                 var updateAlert = downloadUpdateRadioButton.Checked && dataGridViewPatches.SelectedRows.Count > 0
                     ? (XenServerPatchAlert) ((PatchGridViewRow) dataGridViewPatches.SelectedRows[0]).UpdateAlert
                     : selectFromDiskRadioButton.Checked
-                        ? GetAlertFromFileName(fileNameTextBox.Text.ToLowerInvariant())
+                        ? GetAlertFromFileName(isValidFile(unzippedUpdateFilePath) ? unzippedUpdateFilePath.ToLowerInvariant() : fileNameTextBox.Text.ToLowerInvariant())
                         : null;
                 if (updateAlert != null && updateAlert.NewServerVersion != null)
                     return WizardMode.NewVersion;
                 return WizardMode.SingleUpdate;
             } 
         }
 
+        public KeyValuePair<XenServerPatch, string> PatchFromDisk
+        {
+            get {
+                return selectFromDiskRadioButton.Checked && FileFromDiskAlert != null
+                    ? new KeyValuePair<XenServerPatch, string>(FileFromDiskAlert.Patch, SelectedNewPatch)
+                    : new KeyValuePair<XenServerPatch, string>(null, null);

```

Comment:
It may be safer to add null checks in the actions for the patchFromDisk field itself; here it is carefully set to a `<null,null>` pair, but this might change in future (actually I don't see why not set it simply to null here too).

Reply:
KeyValuePair<K,V> can't be null, because it is a struct.

Reply:
Oh indeed - my bad, then please ignore this comment.